### ISSUE TITLE
[FSDP2] Fractional CPU offload policy (PartialOffloadPolicy)

### ISSUE THREAD

**Issue by joemunene-by** [pytorch/pytorch#187615]
## Summary
FSDP2 (`torch.distributed.fsdp.fully_shard`) offers exactly two offload behaviors today through `OffloadPolicy`: keep all sharded parameters on device (the base `OffloadPolicy`), or offload all of them to CPU (`CPUOffloadPolicy`). There is no supported way to offload a fraction of parameters. This issue proposes a `PartialOffloadPolicy(offload_ratio)` to close the gap, and asks for maintainer direction on the API before a PR.

## Motivation
The offload decision is already stored per parameter. Each `FSDPParam` carries an `offload_to_cpu` boolean, set uniformly from the policy at init time. The runtime machinery beneath the policy is therefore finer-grained than the policy surface that drives it. On memory-constrained single-accelerator training (the case where offload matters most), all-or-nothing offload is a blunt instrument: full offload pays host-device copy latency on every parameter even when only a fraction needs to leave the device to fit the step, and no-offload simply OOMs. A ratio lets a user trade exactly as much copy bandwidth as they must to fit, and no more.

This is the natural continuation of the per-parameter FSDP direction (RFC #114299) and is adjacent to the activation-offload work in #174960, on the parameter axis rather than the activation axis.

## Proposed API
```python
@dataclass(frozen=True)
class PartialOffloadPolicy(OffloadPolicy):
    offload_ratio: float = 1.0  # fraction of parameter numel to place on CPU
```
- `offload_ratio == 0.0` reproduces base `OffloadPolicy` (nothing offloaded).
- `offload_ratio == 1.0` reproduces `CPUOffloadPolicy` (everything offloaded).
- In between, a deterministic selector chooses which parameters are offloaded.

## Selector semantics (proposed)
Largest-first prefix-stop, per communication group: sort the group's parameters by numel descending, offload the longest prefix whose cumulative numel does not exceed `floor(offload_ratio * group_total_numel)`, stop at the first parameter that would exceed the budget. This selection is:
- bounded: offloaded numel never exceeds the target fraction,
- monotonic: raising the ratio only extends the prefix, so the offloaded set grows as a superset (no thrashing of which parameters offload),
- deterministic: stable tie-break by parameter index,
- boundary-exact: identical to the two existing policies at 0.0 and 1.0.

## Questions for maintainers
1. Is the parameter-numel-fraction the right knob, or would a memory-byte budget (accounting dtype) be preferred as the primary API?
2. Should the selector live in the policy, in `_fsdp_init`, or in the param group? Draft places the selection at the per-`FSDPParam` offload assignment site.
3. Preference on prefix-stop vs an exact-fill selector, given the monotonicity/superset property argues for prefix-stop?

A reference implementation with a standalone, torch-free unit test for the selector (boundary equivalence, monotonicity, never-overshoot, determinism) is ready and can be opened as a draft PR on request.

cc @weifengpy @zhaojuanmao @mrshenli @rohan-varma @awgu @fegin @chauhang @mori360 @pragupta @ppwwyyxx

**Comment by ezyang** [comment]
also cc @mlazos

**Comment by fegin** [comment]
cc., @weifengpy Any thought on the proposed design?

**Comment by mlazos** [comment]
@joemunene-by the way we handle this in the core memory policies is that we allow the user to pass a policy function, this function is called on each node in the graph to allow a user to mark which nodes should be offloaded.  See here as a starting point: https://github.com/pytorch/pytorch/blob/f3e7e918c66cc2eebb7183ae00c409555c2b24b8/torch/utils/checkpoint.py#L1260-L1261

Perhaps we can do something similar here?

In torchtitan, I implemented a budget parameter and the default is to offload the largest params first. This works pretty well, and this is also an approach I think is reasonable.

**Comment by joemunene-by** [comment]
Thanks, this is a helpful direction. A user-provided callable is a better primary surface than a fixed ratio, since it covers the cases a single scalar cannot: offload by module path, by dtype, or by name pattern.

One difference from the activation policy in `checkpoint.py` is worth noting. That policy runs per op during the forward pass because the set of saved tensors is dynamic. FSDP2 parameter offload is decided once at init, since `offload_to_cpu` is fixed per `FSDPParam` before the first all-gather. So the analog here is a callable invoked once per parameter at construction, given enough metadata to decide (numel, dtype, FQN, owning module), returning whether that parameter offloads. The decision stays static and there is no per-step cost.

On the budget, that matches the selector I proposed (largest-first up to a target), so I am glad it held up in torchtitan. I would fold it in as the default callable rather than a separate policy, so the API is a single callable with a sensible default and an escape hatch for everything else.

A few things I would want your read on before revising the draft:

1. Return type. A plain `bool`, or a small enum mirroring `CheckpointPolicy` so `MUST_*` vs `PREFER_*` can interact with other subsystems later. bool is simpler; the enum leaves forward room.
2. Budget units for the default. Parameter numel is dtype-agnostic and simple, but a byte budget (`numel * itemsize`) is what actually maps to device memory under mixed precision. I lean toward bytes since the user is reasoning about fitting memory.
3. Budget granularity. Largest-first within each communication group (`FSDPParamGroup`), or across the whole model. Per-group keeps the offloaded fraction even across groups; global gives a tighter fit at the cost of uneven per-group copy.

I can rework the draft around a single `offload_policy_fn` with the largest-first byte budget as the default, keeping the standalone selector test. I will open it as a draft PR once you have a preference on the above.

**Comment by weifengpy** [comment]
@joemunene-by thanks for the proposal! cpu offloading decision at parameter-level is plausible! I just learnt the D2H and H2D actually happens at parameter level, because of copy-in to AG flat buffers

I personally want to decouple two things, preferring simple over easy
* specify cpu offloading at parameter-level, I would prefer this to be a {FQN: cpu_offload_policy} 
* which parameter to offload, may it be `offload_ratio` or something else, I prefer the helper function eventually return {FQN: cpu_offload_policy}. I don't mind landing hosting those helper functions into torch FSDP2

what do you think?

**Comment by joemunene-by** [comment]
This decoupling is the right call, and the `{FQN: OffloadPolicy}` map as the core surface is clean: it is data the user can inspect and diff, the mechanism stays free of any selection logic, and the two existing policies fall out as the per-entry values. I am on board with keeping mechanism and policy as separate layers.

Good to have the runtime confirmation that D2H/H2D is per parameter through the copy-in to the all-gather flat buffer. That is what makes the per-FQN map cheap: no new machinery under it, just which entries get a CPU policy.

A few specifics so the PR matches what you have in mind:

1. Value type. I would make the value an existing `OffloadPolicy` instance, so `CPUOffloadPolicy(pin_memory=...)` is expressible per parameter and the current two policies stay the endpoints. FQNs absent from the map default to on device. Is that what you meant by `cpu_offload_policy`, or were you picturing a plain bool?
2. How the map reaches `fully_shard`. I would overload the existing `offload_policy` argument to accept either a single `OffloadPolicy` (today's behavior) or a `Mapping[str, OffloadPolicy]`, rather than add a new keyword. That keeps the call site unchanged for existing users.
3. FQN namespace. I am assuming FQNs are relative to the module passed to `fully_shard`, since the call is per module. Worth pinning down so the helpers resolve names the same way.

For the policy layer, the helpers stay pure: they take the module's named parameters plus a budget or ratio and return the `{FQN: OffloadPolicy}` map, with largest-first budget as the default. That keeps them unit-testable without a distributed context, which is how I already have the selector tested (boundary equivalence at the two endpoints, monotonicity, never-overshoot, determinism).

If that lines up with your intent, I will open a draft PR in two parts: the mechanism change to accept the map, and the largest-first budget helper under `torch.distributed.fsdp` that returns it, with the standalone selector test. Confirm the value type and I will match it.

**Comment by weifengpy** [comment]
> I would make the value an existing OffloadPolicy instance, so CPUOffloadPolicy(pin_memory=...) 

> overload the existing offload_policy argument to accept either a single OffloadPolicy (today's behavior) or a Mapping[str, OffloadPolicy]

> FQN namespace. I am assuming FQNs are relative to the module passed to fully_shard

appreicate your attention to details - all 3 things sounds good

I want to add 2 things
* confirm adam / adamw works for a mixture of cuda and cpu tensors. I don't know if adam(foreach=) or adam(fused=) can be true anymore. it might have to be per-param kernel launch
* raise value error if FQNs does not match any existing parameters

feel free to draft PR - welcome your contribution

**Comment by joemunene-by** [comment]
On the optimizer question, I traced the dispatch. The short version: the foreach and fused step implementations already handle a mixed-device group correctly, but the default auto-selection does not pick them, so the risk you flagged is real for the default config and avoidable by being explicit.

Detail:

- Both `_multi_tensor_adam` (foreach) and `_fused_adam` call `_group_tensors_by_device_and_dtype` and launch one kernel per `(device, dtype)` bucket. A param group split across CPU and CUDA shards becomes two buckets and runs one foreach/fused launch each. So it is per device bucket, not per parameter.
- The catch is `_default_to_fused_or_foreach`, which gates auto-selection on `_get_foreach_kernels_supported_devices()`. That list is `["cuda", "xpu", "mtia", privateuse1]` and excludes CPU. With `foreach=None, fused=None` (the Adam/AdamW default), a group that mixes CPU and CUDA fails the `all(...)` check and falls to `_single_tensor_adam`, the per-parameter loop. That is the cliff.
- `fused` is the clean path here. `_get_fused_kernels_supported_devices()` includes both `cuda` and `cpu`, and the CPU fused Adam/AdamW kernel exists, so `fused=True` stays valid for a mixed group and dispatches a fused kernel per device. Explicit `foreach=True` also works, with the CPU bucket on the slower foreach path.

Net: correctness holds regardless. The only failure mode is silent degradation to single-tensor when the optimizer is left on its default for a mixed group. I will document this on the helper and recommend `fused=True` (or explicit `foreach=True`) for partial offload, rather than have FSDP override the user's optimizer flags.

On validation, I will raise `ValueError` for any FQN in the map that does not resolve to a parameter of the `fully_shard`-ed module, listing the offending keys.

I will open the draft PR along these lines: `offload_policy` accepting `Mapping[str, OffloadPolicy]`, the largest-first budget helper under `torch.distributed.fsdp` returning the map, the FQN check, and the standalone selector test. I will tag you on it.

**Comment by joemunene-by** [comment]
Drafted in #188120.

It follows the mechanism/policy split: `offload_policy` now accepts a `Mapping[str, OffloadPolicy]` keyed by FQN, and `cpu_offload_by_budget(module, budget_bytes)` returns that map, largest parameters first. The two points from your last comment are handled: an unknown FQN raises `ValueError`, and the optimizer behavior is documented on the helper, namely that `fused=True` (or explicit `foreach=True`) keeps the per-device multi-tensor step while the default auto-selection drops a mixed group to the single-tensor path.

Still a draft. Distributed tests and docs to follow once the surface looks right to you.
